### PR TITLE
Add merge index image endpoint

### PR DESCRIPTION
**Pull request by querti**
- The stages were implemented according to the design doc
- '_prepare_request_for_build' function has been updated to
  accomodate the new request type.
- Deprecation based on the image labels was implemented as well,
  although it will currently not be used.
- Stages can be skipped if it's discovered that they are a no-op.
- Intermediate index images have to be uploaded to the registry as
  opm cannot work with local images.
- One thing which hasn't been implemented is the propagation of
  default channel deletion error. The reason is that the error
  format has not yet been agreed upon.
- Added new endpoint, along with new relational models specification,
  updated data validation, database migration, and documentation.
- Pull specifications of the source and target index images are
  updated during the course of the task.

Unfortunately, this PR contains a commit from #137. It requires functions introduced by that PR, which hasn't yet been merged. I will rebase the new changes on top of master once #137 is merged.

I should mention that I haven't successfully tested overwrite_from_index, because quay always gives error "manifest invalid: manifest invalid". Has anyone encountered this?

### CODE REVIEW

**Comment by querti** [comment]
Data I used for testing:
`curl -u: --negotiate -X POST -H "Content-Type: application/json"  http://localhost:8080/api/v1/builds/merge-index-image -d \
    '{
       "source_from_index": "quay.io/lgallovi/iib-test:4.5.10",
       "target_from_index": "quay.io/lgallovi/iib-test:4.6.4",
       "deprecation_list": ["registry-proxy.engineering.redhat.com/rh-osbs/quay-quay-container-security-operator-metadata@sha256:5c089eb32a2ba03e8901f4e4f55b5b0724b4629ab79aa71ba392b02c3c8c3b52"],
       "binary_image": "registry-proxy.engineering.redhat.com/rh-osbs/openshift-ose-operator-registry:v4.6"
     }'`

**Comment by querti** [comment]
@yashvardhannanavati @mprahl @lcarva @lipoja @midnightercz @chandwanitulsi review pls

**Review by querti — COMMENTED** [review]
After #145 is merged, I have to incorporate the changes

File: iib/workers/tasks/build.py
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:652]: Could you please change target_from_index to target_index everywhere?

File: iib/workers/tasks/build.py
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:677]: This doesn't seem right, the source_from_index data will be added here and a few more. Could you please reword this?

File: iib/workers/tasks/build.py
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:722]: This code seems to be duplicated in the same function. How about creating a helper method?

File: iib/workers/tasks/build.py
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:797]: Same, could you please rename `target_from_index_resolved` to `target_index_resolved`?

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1082]: Could you please rename target_from_index to target_index here too?

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1093]: for the merge-index-image endpoint?

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1102]: Why do you need the bundles to be specified with their hash? Doesn't IIB already resolve them to their hases? You could reuse that method

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py:379]: This should be discussed in #137. Although I see where you're coming from, the code is too intertwined to be easily separable.

File: iib/workers/tasks/utils.py
Comment (querti) [reply iib/workers/tasks/utils.py:307]: This should be discussed in #137. Reading output from a running process is blocking, and since there's no deterministic way to know when to stop reading, the read would end up hanging forever. I've actually tried many different solutions for non-blocking reading, and this was the only one that worked reliably.

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py:1529]: This is a part of the code that will be uncommented later. Right now, it's not necessary to rebuild the intermediate index image, so it's commented.

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py:1533]: I made the message a bit more detailed

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: All occurrences should be changed

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: Hopefully it's more accurate now.

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: Hmm it's only two lines that are duplicated everywhere. And they are both variable value setters. Does it really make sense to create a helper function?

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: done

File: iib/web/models.py
Comment (querti) [reply iib/web/models.py]: done

File: iib/web/models.py
Comment (querti) [reply iib/web/models.py]: fixed.

File: iib/web/models.py
Comment (querti) [reply iib/web/models.py:1102]: I think I saw the requirement for the bundles to be specified by a digest in the design doc. "It’ll be a yaml file with package name as the key and the bundle pull_spec list as value specified with digest. " I guess the data validation doesn't have to be done on IIB's side, but should instead be handled by the script. Should I changed it to accept tags as well?

**Comment by querti** [comment]
@midnightercz @yashvardhannanavati @lcarva @mprahl @chandwanitulsi @lipoja check pls

**Comment by yashvardhannanavati** [comment]
@querti Could you please rebase this PR?

File: iib/web/api_v1.py
Comment (yashvardhannanavati) [inline iib/web/api_v1.py:660]: Why not simply call _get_safe_args here?

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1110]: f string is not required here

File: iib/web/models.py
Comment (yashvardhannanavati) [reply iib/web/models.py:1107]: I think all you need to check here is that bundles is an array of strings. Let the bundle resolution in the code take care of valid/invalid bundles.

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1105]: Let's use `deprecation_list` as the variable name instead of bundles to avoid confusion.

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/workers/tasks/build.py
Comment (yashvardhannanavati) [inline iib/workers/tasks/build.py:1121]: `skopeo_inspect` is an expensive operation. Since you're simply using a single label, could you use the get_image_label method to get it? (It uses the --config option which is extremely fast). Check out https://github.com/release-engineering/iib/commit/f58767585bfedbbb4d771c9f238d23d27a1aeaf6 for reference. If you're using it in any other place, could you please replace those too?

File: iib/workers/tasks/build.py
Comment (yashvardhannanavati) [reply iib/workers/tasks/build.py:1541]: I'm not a fan of keeping non-functional commented code. Could you please remove it and store it in your branch or a snippet?

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/web/api_v1.py
Comment (querti) [reply iib/web/api_v1.py]: Fixed

File: iib/web/models.py
Comment (querti) [reply iib/web/models.py]: Fixed

File: iib/web/models.py
Comment (querti) [reply iib/web/models.py]: The reason why I added it was to check if each bundle is specified via digest, not to verify bundle validity.

I can still remove it if you'd like.

File: iib/web/models.py
Comment (querti) [reply iib/web/models.py]: Fixed

File: iib/web/models.py
Comment (querti) [reply iib/web/models.py]: Fixed

File: iib/web/static/api_v1.yaml
Comment (querti) [reply iib/web/static/api_v1.yaml]: Fixed

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: Changed it for now, but will wait for confirmation

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py:716]: I added a comment to the API docs. I guess the other option would be to use two tokens, but that seems a bit too contrived and doesn't seem like our use-case.

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: You're right. Changed.

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: Changed.

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: Added.

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: Fixed.

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: Good catch. Fixed.

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: Sorry, I don't think I understand. I thought that currently this type of deprecation should not be used at all. '_get_deprecated_bundles_by_ocp_version' is not currently used anywhere in the code.

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: Fixed.

File: iib/workers/tasks/build_merge_index_image.py
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_merge_index_image.py:58]: Why not call build_image here for every arch present? Check out handle_add_request for reference.

File: iib/workers/tasks/build_merge_index_image.py
Comment (yashvardhannanavati) [inline iib/workers/tasks/build_merge_index_image.py:317]: Instead of calling opm_index_add twice in this function and in _create_intermediate_index_image, what's stopping us from finding the missing bundles first and then simply calling opm_index_add once along with the missing bundles to form the intermediate index image. Then simply call build_image once for each arch on the index image. Wouldn't this approach reduce complexity?

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: iib/workers/tasks/build_merge_index_image.py
Comment (querti) [reply iib/workers/tasks/build_merge_index_image.py]: I thought that it would be more efficient to built only one arch for the intermediate stages. All archs are build for the final index image. Was that a bad thinking on my part?

File: iib/workers/tasks/build_merge_index_image.py
Comment (querti) [reply iib/workers/tasks/build_merge_index_image.py:286]: Yeah, I noticed that the process could be optimized, but decided to implement it as it was described in the design doc in case I was missing some information.

I have simplified the code according to your suggestion.

**Review by midnightercz — CHANGES_REQUESTED** [review]

File: iib/workers/tasks/build_merge_index_image.py
Comment (querti) [reply iib/workers/tasks/build_merge_index_image.py]: Whoops, you're right. Fixed and updated tests to catch it.

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py:733]: @yashvardhannanavati thoughts?

File: iib/workers/tasks/build_merge_index_image.py
Comment (querti) [reply iib/workers/tasks/build_merge_index_image.py:69]: IIB Already performs a similar operation when checking for already present bundles, and it doesn't display any warnings.
I suppose we could, but it might become very spammy.

If such a scenario does happen, it would be pretty big trouble, as IIB won't be guaranteed to find all the missing bundles

Comment (querti) [reply iib/workers/tasks/build_merge_index_image.py:69]: @yashvardhannanavati  Thoughts?

File: iib/workers/tasks/build.py
Comment (shawn-hurley) [reply iib/workers/tasks/build.py:733]: I think we probably want to point to n+1 of OCP. I agree making this configurable is desirable, I think we have two configuration stories that we probably want to make:

1. make ocp_version configurable (the label and the default value)
2. make distribution_scoep configurable (the label and the default value) 

at least that is my opinion

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py]: I removed the label deprecation logic, as discussed in the meeting.

File: iib/workers/tasks/build_merge_index_image.py
Comment (querti) [reply iib/workers/tasks/build_merge_index_image.py:69]: I added an error message if this happens.

File: iib/workers/tasks/build.py
Comment (querti) [reply iib/workers/tasks/build.py:733]: Hmm so what do I do here?

**Comment by yashvardhannanavati** [comment]
Closing this in favor of #162

**Pull request closed without merging**
